File: binary_tree/count_good_nodes_in_binary_tree/solution2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func goodNodes(root *TreeNode) int {
    var (
		dfs func(node *TreeNode, v int)
		ret int
	)

	dfs = func(node *TreeNode, v int) {
		if node == nil {
			return
		}

		if node.Val >= v {
			v, ret = node.Val, ret+1
		}
		dfs(node.Left, v)
		dfs(node.Right, v)
	}
	dfs(root, math.MinInt)

	return ret
}

/*
	执行用时分布：0ms，击败 100.00%
	消耗内存分布：14.20MB，击败 35.20%
*/

File: binary_tree/validate_binary_search_tree/solution2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

min_val, max_val = -2**63, 2**63

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def check(root: Optional[TreeNode]) -> tuple[bool, int, int]:
            if root is None:
                return True, max_val, min_val

            lok, lmin, lmax = check(root.left)
            rok, rmin, rmax = check(root.right)

            return lok and rok and lmax < root.val and root.val < rmin, lmin if root.left is not None else root.val , rmax if root.right is not None else root.val

        ans, _, _ = check(root)
        return ans

# 执行用时分布：4ms，击败 17.02%
# 消耗内存分布：20.53MB，击败 85.52%